### ISSUE TITLE
Return files destination uris in all ToGCS operators

### ISSUE THREAD

**Issue by FHoffmannCode** [apache/airflow#11323]
**Description**

Currently some storage operators return files destination uris list for example `GoogleSheetsToGCSOperator`. It would be a good idea to make all of such operators to return this list.

**Use case / motivation**

1. Having list of destination uris for files transferred with operator is useful for debugging and troubleshooting
2. Inconsistent operator behaviour might be confusing for users.

**Edit (by @shahar1 on Jan. 31, 2026)**
For operators that currently return a single entity, add the flag `unwrap_single` which controls whether the returned value is either `list[obj]` or `obj` (usually `obj` is `str` but could also be a `dict`). Until we apply consistency for all operators (`unwrap_single=False`), **which will be a breaking change** - set for now the default to return the single entity (`unwrap_single=True`).

## Tracking (added by @shahar1 on Jan. 24, 2026)

All operators should return `list[str]` of destination URIs (`gs://bucket/object`).

### To Do (Generated by AI, please double check if current situation is as described)

- [x] **ADLSToGCSOperator** (`transfers/adls_to_gcs.py`)
  - Current: `list[str]` (ADLS **source** paths) → Need: destination URIs
  - Assignee: @Abhishekmishra2808  |  PR: #61463 

- [x] **AzureFileShareToGCSOperator** (`transfers/azure_fileshare_to_gcs.py`)
  - Current: `list[str]` (**source** filenames) → Need: destination URIs
  - Assignee: @nailo2c  |  PR: #61048

- [x] **AzureBlobStorageToGCSOperator** (`transfers/azure_blob_to_gcs.py`)
  - Current: `str` → Need: `list[str]` (wrap in list)
  - Assignee: @nailo2c  |  PR: #61048

- [x] **GCSToGCSOperator** (`transfers/gcs_to_gcs.py`)
  - Current: `None` → Need: destination URIs
  - Assignee: @yuseok89  |  PR: #61320

- [x] **LocalFilesystemToGCSOperator** (`transfers/local_to_gcs.py`)
  - Current: `None` → Need: destination URIs
  - Assignee: @yuseok89  |  PR: #61355

- [x] **GoogleDriveToGCSOperator** (`transfers/gdrive_to_gcs.py`)
  - Current: Inconsistent → Need: destination URIs
  - Assignee: @Ajay9704  |  PR: https://github.com/apache/airflow/pull/61347

- [x] **HttpToGCSOperator** (`transfers/http_to_gcs.py`)
  - Current: `None` → Need: destination URIs
  - Assignee: @Srabasti  |  PR: https://github.com/apache/airflow/pull/61306

- [x] **FacebookAdsReportToGcsOperator** (`transfers/facebook_ads_to_gcs.py`)
  - Current: `None` → Need: destination URIs
  - Assignee: @Abhishekmishra2808  |  PR: : #61349

- [x] **GoogleCalendarToGCSOperator** (`transfers/calendar_to_gcs.py`)
  - Current: object name (not full URI) → Need: full `gs://` URIs
  - Assignee: @Ajay9704  |  PR: https://github.com/apache/airflow/pull/61284

- [x] **GoogleSheetsToGCSOperator** (`transfers/sheets_to_gcs.py`)
  - Current: `list[str]` (object names only) → Need: full `gs://` URIs
  - Assignee: @Ajay9704   |  PR: https://github.com/apache/airflow/pull/61347

- [ ] **SalesforceToGcsOperator** (`transfers/salesforce_to_gcs.py`)
  - Current: `str` → Need: `list[str]` (wrap in list)
  - Assignee: @Prab-27  |  PR: https://github.com/apache/airflow/pull/61659

- [x] **GoogleAdsToGcsOperator** (`ads/transfers/ads_to_gcs.py`)
  - Current: `None` → Need: destination URIs
  - Assignee: @Abhishekmishra2808  |  PR: https://github.com/apache/airflow/pull/61334

- [x] **GoogleDisplayVideo360SDFtoGCSOperator** (`marketing_platform/operators/display_video.py`)
  - Current: non-list / non-URI → Need: full `gs://` URIs
  - Assignee: @Ajay9704 | PR: https://github.com/apache/airflow/pull/61709

- [x] **S3ToGcsOperator** (`cloud/transfers/s3_to_gcs.py‎`)
  - Need: adjustments to deferrable mode. 
  - Assignee: @AlexisBRENON | PR: https://github.com/apache/airflow/pull/49768

**Comment by eladkal** [comment]
This issue spawn from https://github.com/apache/airflow/pull/10991#discussion_r500837769
It probably can be useful but it's also a breaking change and so far I didn't see users complaining about it.
Could you please clarify the TODO part on this issue?
What operators need adjustments? Is it localized to GCP or also other providers?

**Comment by uranusjr** [comment]
Yeah I don’t see how this could break user code in practice. Should be doable.

**Comment by potiuk** [comment]
Yeah. Nice feature to add. Would you like to implement it @FHoffmannCode ? Should I assign it to you ?

**Comment by eladkal** [comment]
> Yeah I don’t see how this could break user code in practice.

The suggestion was to change the returned value to make operators consistent. If changed then the value pushed to xcom is not as users are expecting.

**Comment by uranusjr** [comment]
It depends on what are correctly being returned; I’m assuming most (if not all) are currently returning None, which won’t have backward compatibility issues (except if someone is depending on an XCom is _not_ pushed, but why would anyone do that). Are there instances of an operator currently pushing something else?

**Comment by potiuk** [comment]
Yeah. Same here. I think it's no harm to return "more".

**Comment by shahar1** [comment]
@eladkal Please assign me

**Comment by potiuk** [comment]
assigned

**Comment by aaron-y-chen** [comment]
I'll take azure related tasks :)

**Comment by shahar1** [comment]
Thanks!
General comment - when standardizing an output, be **careful** not to break existing behavior without early notice. In some cases it should be even controlled by a flag.
For example, when changing a string (`"gs://uri1"`) to a list of strings `["gs//uri1"]` - this is a breaking change, and it should be controlled (by default it should show current behavior - if it needs to be changed, then also add a deprecation warning).

**Comment by shahar1** [comment]
> Thanks! General comment - when standardizing an output, be **careful** not to break existing behavior without early notice. In some cases it should be even controlled by a flag. For example, when changing a string (`"gs://uri1"`) to a list of strings `["gs//uri1"]` - this is a breaking change, and it should be controlled (by default it should show current behavior - if it needs to be changed, then also add a deprecation warning).

Added a note regarding the flag in the issue's description (`unwrap_single`)

**Comment by yuseok89** [comment]
I'll take `GCSToGCSOperator` task.

**Comment by yuseok89** [comment]
Also, I’d like to take the `LocalFilesystemToGCSOperator` task as well.

**Comment by Prab-27** [comment]
@shahar1 ,I'd like to work on `SalesforceToGcsOperator `

**Comment by Srabasti** [comment]
Thanks for sharing @shahar1 ! Can you please assign me HttpToGCSOperator?

**Comment by ahmadtfarhan** [comment]
> I'll take azure related tasks :)

Same

**Comment by shahar1** [comment]
> > I'll take azure related tasks :)
> 
> Same

It seems that both `ToGCS` operators were handled in #61048.
Would you like to take other operators, or rather make a similar effort in the other direction, i.e., `ToAzureBlob`?
For the latter I'd prefer to handle it in a separate issue though.

**Comment by Ajay9704** [comment]
@shahar1 , I am willing to work on the two other operators:

GoogleSheetsToGCSOperator - Update to return full gs:// URIs with unwrap_single flag
GoogleDriveToGCSOperator - Update to return full GCS URIs with unwrap_single flag

I've analyzed their current implementations and am ready to implement the same patterns we used for the previous operators (GoogleCalendarToGCSOperator and GoogleDisplayVideo360SDFtoGCSOperator).
If these operators are still unassigned, I'm ready to take them on

**Comment by ahmadtfarhan** [comment]
> > > I'll take azure related tasks :)
> > 
> > 
> > Same
> 
> It seems that both `ToGCS` operators were handled in [#61048](https://github.com/apache/airflow/pull/61048). Would you like to take other operators, or rather make a similar effort in the other direction, i.e., `ToAzureBlob`? For the latter I'd prefer to handle it in a separate issue though.

There is only one with `ToAzureBlob`: `S3ToAzureBlobStorageOperator` 
should that also be the same for `ToWasb`?

**Comment by shahar1** [comment]
> > > > I'll take azure related tasks :)
> > > 
> > > 
> > > Same
> > 
> > 
> > It seems that both `ToGCS` operators were handled in [#61048](https://github.com/apache/airflow/pull/61048). Would you like to take other operators, or rather make a similar effort in the other direction, i.e., `ToAzureBlob`? For the latter I'd prefer to handle it in a separate issue though.
> 
> There is only one with `ToAzureBlob`: `S3ToAzureBlobStorageOperator` should that also be the same for `ToWasb`?

Yeah, I guess that it could be considered as well.
Would you like to create a separate issue for that, or even the PRs directly? 
I'll be happy to review when applicable.

**Comment by Abhishekmishra2808** [comment]
I’d like to take up `GoogleAdsToGcsOperator` (`ads/transfers/ads_to_gcs.py`) from this issue.

Plan:
- Update the operator to return `List[str]` of destination GCS URIs instead of `None`
- Follow the return-value pattern used by other updated `*ToGCSOperator`s
- Add/update unit tests to validate the returned GCS URIs
- Update docstrings/type hints if required

Please assign it to me if this works. Thanks!

**Comment by Abhishekmishra2808** [comment]
Hi @shahar1 , I’d like to work on this issue `FacebookAdsReportToGcsOperator`. Could you please assign it to me? Thanks!

**Comment by molcay** [comment]
Hi @shahar1,
I have a concern about this implementation.
Since the `XCom` has a hard limit to 1GB (due to the DB field type `JSONB`) and the return value from a Operator stored in the `XCom`, do we need to do something extra? I know the 1GB is a huge size but still... 

Lets assume you are migration a bucket to another bucket (this can be trained AI model data) and try to move the data between buckets. I think we can hit that limit in this kind of scenario

**Comment by shahar1** [comment]
> Hi @shahar1,
> I have a concern about this implementation.
> Since the `XCom` has a hard limit to 1GB (due to the DB field type `JSONB`) and the return value from a Operator stored in the `XCom`, do we need to do something extra? I know the 1GB is a huge size but still... 
> 
> Lets assume you are migration a bucket to another bucket (this can be trained AI model data) and try to move the data between buckets. I think we can hit that limit in this kind of scenario

Hey, thanks for raising these concerns!


1. Some operators included in this issue (still) transfer only one file, so I wouldn't worry about them.

2. Specificaly for transferring an entire bucket, especially if it includes a lot of files, I don't think that that the "regular" GCS2GCS transfer operator is right tool to use (using Airflow's resources for transferring), but I'd rather use the GCP's storage transfer service instead (deferring to GCP).

3. If you insist to use the GCS2GCS operator, you could always pass `do_xcom_push = False` and avoid the xcom being pushed at all :)

**Comment by molcay** [comment]
> > Hi [@shahar1](https://github.com/shahar1),
> > I have a concern about this implementation.
> > Since the `XCom` has a hard limit to 1GB (due to the DB field type `JSONB`) and the return value from a Operator stored in the `XCom`, do we need to do something extra? I know the 1GB is a huge size but still...
> > Lets assume you are migration a bucket to another bucket (this can be trained AI model data) and try to move the data between buckets. I think we can hit that limit in this kind of scenario
> 
> Hey, thanks for raising these concerns!
> 
> 1. Some operators included in this issue (still) transfer only one file, so I wouldn't worry about them.
> 2. Specificaly for transferring an entire bucket, especially if it includes a lot of files, I don't think that that the "regular" GCS2GCS transfer operator is right tool to use (using Airflow's resources for transferring), but I'd rather use the GCP's storage transfer service instead (deferring to GCP).
> 3. If you insist to use the GCS2GCS operator, you could always pass `do_xcom_push = False` and avoid the xcom being pushed at all :)

Hey, thank you for the answer 🙇🏼 

I was concerned about if someone is using bucket-to-bucket transfer via "bad practice" :)
I was not aware of the `do_xcom_push` flag. If we have some way to to avoid this, it is good 👍🏼 

Currently, from the user perspective, if someone is using this "bad practice", they will face this issue. After that, they have some options:
- Change the implementation and use the proper way (like Storage Transfer Service)
- Change the implementation and pass `do_xcom_push` to avoid this behavior

> I am not sure there is a service for cross cloud transfers however we already have the `do_xcom_push`.

**Comment by Abhishekmishra2808** [comment]
Hey @shahar1 ,
I’d like to take up `ADLSToGCSOperator` (transfers/adls_to_gcs.py) from this issue.

Plan:
- Update the operator to return list[str] of destination GCS URIs
- Follow the return-value pattern used by other updated *ToGCSOperators
- Add/update unit tests to validate returned URIs
- Update docstrings/type hints if needed

Please assign this to me if it works. Thanks!

**Comment by yuseok89** [comment]
There are cases where extreme misuse isn't defended against, and defensive logic and exception branches can make the code harder to maintain.
Still, users who are already using the operator in a "misuse" way (e.g. bucket-to-bucket transfer with many files) could suddenly run into the XCom limit once we start returning destination URIs. In my view, if those users are to be considered, something like a parameter named `return_destination_uris` with a default of `False` could be one option. It would keep the default safer (no large return value) while still allowing callers who need the list to opt in.

**Comment by shahar1** [comment]
> There are cases where extreme misuse isn't defended against, and defensive logic and exception branches can make the code harder to maintain. Still, users who are already using the operator in a "misuse" way (e.g. bucket-to-bucket transfer with many files) could suddenly run into the XCom limit once we start returning destination URIs. In my view, if those users are to be considered, something like a parameter named `return_destination_uris` with a default of `False` could be one option. It would keep the default safer (no large return value) while still allowing callers who need the list to opt in.

Personally, I don't think that we should be overly gentle with extreme misuse. XComs are inherently a metadata mechanism, and if the list of files is overly large - the users should adopt the appropriate patterns (e.g., using a custom XCom backend or migrating to Google-cloud based transfer operation). As I mentioned, the option to avoid pushing to xcoms already exists, and we're already do our best to implement it in a slow backward-comaptibile way. We shouldn’t design defaults around anti-patterns that go against Airflow best practices.

**Comment by yuseok89** [comment]
@shahar1
I have some bandwidth and can help more on this issue.
Feel free to assign me anything that still needs help.

**Comment by shahar1** [comment]
> [@shahar1](https://github.com/shahar1) I have some bandwidth and can help more on this issue. Feel free to assign me anything that still needs help.

Thank you! It's seems that all other operators are currently assigned.
**However,** if you could somehow help with running the system tests for the other operators as well (similar to what you did in your PRs) - it will be **very** helpful, as we'll be able to merge them quicker.
If you're unnable to do it, you're welcome to tackle other GitHub issues.

**Comment by yuseok89** [comment]
> Thank you! It's seems that all other operators are currently assigned.
> **However,** if you could somehow help with running the system tests for the other operators as well (similar to what you did in your PRs) - it will be **very** helpful, as we'll be able to merge them quicker.
> If you're unnable to do it, you're welcome to tackle other GitHub issues.

Thanks! Yes, I can do that. If running the system tests helps speed up merges, I'm happy to help as much as I can. Let me know if there's anything else you'd like me to help with.

**Comment by shahar1** [comment]
Quick update - the entire list is populated, and at least 4 PRs are merged and expected to be included in the upcoming release.
We need some help with running the system tests for some of the open PRs.
Thanks everyone for your hard work and dedication so far, keep up!

@yuseok89 @nailo2c @Ajay9704 @Prab-27 @Srabasti @Abhishekmishra2808

**Comment by potiuk** [comment]
@Ajay9704 We are unassigning you from this issue as part of our updated [assignment policy](https://github.com/apache/airflow/blob/main/contributing-docs/04_how_to_contribute.rst#contribute-code-changes).

This is not meant to discourage your contribution — quite the opposite! You are still very welcome to work on this issue and submit a PR for it. Simply comment that you are working on it and open a PR when ready.

We found that formal assignments were not working well, as they often prevented others from contributing when the assignee was not actively working on the issue.

**Comment by potiuk** [comment]
@yuseok89 We are unassigning you from this issue as part of our updated [assignment policy](https://github.com/apache/airflow/blob/main/contributing-docs/04_how_to_contribute.rst#contribute-code-changes).

This is not meant to discourage your contribution — quite the opposite! You are still very welcome to work on this issue and submit a PR for it. Simply comment that you are working on it and open a PR when ready.

We found that formal assignments were not working well, as they often prevented others from contributing when the assignee was not actively working on the issue.

**Comment by potiuk** [comment]
@nailo2c We are unassigning you from this issue as part of our updated [assignment policy](https://github.com/apache/airflow/blob/main/contributing-docs/04_how_to_contribute.rst#contribute-code-changes).

This is not meant to discourage your contribution — quite the opposite! You are still very welcome to work on this issue and submit a PR for it. Simply comment that you are working on it and open a PR when ready.

We found that formal assignments were not working well, as they often prevented others from contributing when the assignee was not actively working on the issue.

**Comment by potiuk** [comment]
@Abhishekmishra2808 We are unassigning you from this issue as part of our updated [assignment policy](https://github.com/apache/airflow/blob/main/contributing-docs/04_how_to_contribute.rst#contribute-code-changes).

This is not meant to discourage your contribution — quite the opposite! You are still very welcome to work on this issue and submit a PR for it. Simply comment that you are working on it and open a PR when ready.

We found that formal assignments were not working well, as they often prevented others from contributing when the assignee was not actively working on the issue.

**Comment by potiuk** [comment]
@Prab-27 We are unassigning you from this issue as part of our updated [assignment policy](https://github.com/apache/airflow/blob/main/contributing-docs/04_how_to_contribute.rst#contribute-code-changes).

This is not meant to discourage your contribution — quite the opposite! You are still very welcome to work on this issue and submit a PR for it. Simply comment that you are working on it and open a PR when ready.

We found that formal assignments were not working well, as they often prevented others from contributing when the assignee was not actively working on the issue.

**Comment by potiuk** [comment]
@Srabasti We are unassigning you from this issue as part of our updated [assignment policy](https://github.com/apache/airflow/blob/main/contributing-docs/04_how_to_contribute.rst#contribute-code-changes).

This is not meant to discourage your contribution — quite the opposite! You are still very welcome to work on this issue and submit a PR for it. Simply comment that you are working on it and open a PR when ready.

We found that formal assignments were not working well, as they often prevented others from contributing when the assignee was not actively working on the issue.

**Comment by shahar1** [comment]
Closing this issue, as there's only a single operator left which will be treated separately.
Thanks again everyone!

**Comment by yuseok89** [comment]
This issue was open for a long time, and it's great to see it finally closed. Thanks to everyone who contributed, and special thanks to @shahar1 for driving this to completion.

**Comment by nailo2c** [comment]
Thanks for organizing this issue, great job @shahar1!
